globus_sdk/auth/identity_map.py

```diff
@@ -0,0 +1,222 @@
+    improve performance.
+
+    The ``IdentityMap`` will cache its results so that repeated lookups of the same Identity
+    will not repeat work. It will also map identities both by ID and by Username,
+    regardless of how they're initially looked up.
+
+    .. warning::
+
+        If an Identity is not found in Globus Auth, it will trigger a KeyError when
+        looked up. Your code must be ready to handle KeyErrors when doing a lookup.
+
+    Correct usage looks something like so::
+
+        ac = globus_sdk.AuthClient(...)
+        idmap = globus_sdk.ext.IdentityMap(
+            ac, ["foo@globusid.org", "bar@uchicago.edu"]
+        )
+        idmap.add("baz@xsede.org")
+        # adding by ID is also valid
+        idmap.add("c699d42e-d274-11e5-bf75-1fc5bf53bb24")
+        # map ID to username
+        assert (
+            idmap["c699d42e-d274-11e5-bf75-1fc5bf53bb24"]["username"]
+            == "go@globusid.org"
+        )
+        # map username to ID
+        assert (
+            idmap["go@globusid.org"]["id"]
+            == "c699d42e-d274-11e5-bf75-1fc5bf53bb24"
+        )
+
+    And simple handling of errors::
+
+        try:
+            record = idmap["no-such-valid-id@example.org"]
+        except KeyError:
+            username = "NO_SUCH_IDENTITY"
+        else:
+            username = record["username"]
+
+    or you may achieve this by using the :py:meth:`~.IdentityMap.get` method::
+
+        # internally handles the KeyError and returns the default value
+        record = idmap.get("no-such-valid-id@example.org", None)
+        username = record["username"] if record is not None else "NO_SUCH_IDENTITY"
+
+    **Parameters**
+
+    ``auth_client`` (*globus_sdk.AuthClient*)
+      The ``AuthClient`` object which will be used for lookups against Globus Auth
+
+    ``identity_ids`` (*iterable of strings*)
+      A list or other iterable of usernames or identity IDs (potentially mixed together)
+      which will be used to seed the ``IdentityMap`` 's tracking of unresolved Identities.
+
+    ``id_batch_size`` (*int*)
+      A non-default batch size to use when communicating with Globus Auth. Leaving this
+      set to the default is strongly recommended.
+
+    **Methods**
+
+    *  :py:meth:`~.IdentityMap.add`
+    *  :py:meth:`~.IdentityMap.get`
+    *  :py:meth:`~.IdentityMap.__getitem__`
+    *  :py:meth:`~.IdentityMap.__delitem__`
+    """  # noqa
+
+    _default_id_batch_size = 100
+
+    def __init__(self, auth_client, identity_ids=None, id_batch_size=None):
+        self.auth_client = auth_client
+        self.id_batch_size = id_batch_size or self._default_id_batch_size
+
+        # uniquify, copy, and split into IDs vs usernames
+        self.unresolved_ids, self.unresolved_usernames = split_ids_and_usernames(
+            [] if identity_ids is None else identity_ids
+        )
+
+        # the cache is a dict mapping IDs and Usernames
+        self._cache = {}
```

Comment:
Should we consider sharing the cache between instances? I know it likely won't ever apply for CLI usage but since this is in the SDK now...

Reply:
I think the way I would approach that, if anyone ever wants it, is to let you pass a `cache` into the constructor, and default to constructing a new dict.
So we could, someday, have
```python
cache = {}
idmap1 = globus_sdk.IdentityMap(authclient, cache=cache)
idmap2 = globus_sdk.IdentityMap(authclient, cache=cache)
```

The advantage of that is that it lets us support other mappings, like

```python
cache = build_memcache_backed_id_cache()
idmap = globus_sdk.IdentityMap(authclient, cache=cache)
```

but I thought that was all too fancy / unnecessary for v1. Would you like me to take a crack at it, and see if it's simple enough to include?

Reply:
I was thinking of a middle ground of simply making the cache a module level variable because I think it's unlikely that the casual user of this SDK is going to make advanced use of being able to inject their own custom cache.

Reply:
I really don't like the idea of anything which is going to persist over the lifetime of the interpreter.

If someone builds a long-running application and doesn't know that there's a persistent cache underneath, they could get stale results and know that it's happening. It also means that memory usage for the cache will grow over time.

Maybe it's best to just leave this as-is?

Reply:
I'm fine with leaving it as is. If anyone else ends up using this and wants to optimize the caching behavior I suppose they're free to implement a singleton on top of this to get it done.